app/src/main/java/com/cdio/solitaire/controller/GameStateController.kt

```
@@ -0,0 +1,316 @@
+        else if (stack.tail!!.rank.ordinal == 0) stack.tail
+        else null
+    }
+
+    /**
+     * moveStack moves a stack and adds cardToUpdate information to the move object.
+     */
+    private fun moveStack( // TODO: Does this actually need cardsToMove? See popStack comment.
+        sourceStack: CardStack,
+        cardsToMove: Int,
+        targetStack: CardStack
+    ): Card? {
+        targetStack.pushStack(sourceStack.popStack(cardsToMove)!!)
+        return cardToUpdate(sourceStack)
+    }
+
+    private fun moveCard(sourceStack: CardStack, targetStack: CardStack): Card? {
+        return moveStack(sourceStack, 1, targetStack)
+    }
+
+    /**
+     * performMove is a fork function on moveType of the move object.
+     * It maps each moveType to their respective methods.
+     */
+    fun performMove(move: Move) {
+        var cardToUpdate: Card? = null
+        when (move.moveType) {
+            MoveType.MOVE_STACK -> cardToUpdate = moveStack(
+                move.sourceStack!!,
+                getCardPosition(move.sourceCard, move.sourceStack),
+                move.targetStack!!
+            )
+            MoveType.MOVE_FROM_FOUNDATION,
+            MoveType.MOVE_FROM_TALON,
+            MoveType.MOVE_TO_FOUNDATION -> cardToUpdate =
+                moveCard(move.sourceStack!!, move.targetStack!!)
+            MoveType.FLIP_TALON -> flipTalon()
+            MoveType.DRAW_STOCK -> cardToUpdate = drawFromStock()
+            else -> Log.e("MoveTypeNotDefined", "This cannot happen.")
+        }
+        move.cardToUpdate = cardToUpdate
+        gameState.moves.add(move)
+    }
+
+    /**
+     * Searches for the position of a card in a given cardStack, measured from the tail element.
+     */
+    private fun getCardPosition(card: Card?, stack: CardStack): Int {
+        if (card!!.stackID != stack.stackID) {
+            Log.e("CardNotInStack", "Card stackID differs from stack stackID")
+            return -1
```

Consider throwing an exception here instead of just logging the error, since a stack trace maybe could provide a lot of useful information.

Then the caller of this function doesn't have to worry about the returned position being invalid, either.